File: app/planes.py
"""Pure 3-plane resolution logic (arch §3.1/§3.2, AC-4/5/6/6a).

Decides which of the three hard-isolated request planes a request belongs to,
purely from ``Host`` + ``path`` — **no DB, no ASGI** — so it is trivially
unit-testable. ``PlaneDispatchMiddleware`` (``app/middleware.py``) applies it.

Planes:
  * ``mock``  (P1) — the public wildcard mock surface ``<token>.<MOCK_DOMAIN>/<path>``
                     and the localhost path-fallback ``/e/<token>/<path>``. On a
                     mock host **everything** (incl. ``/api`` and ``/static``) is the
                     mock's own path — management is unreachable there by construction.
  * ``api``   (P2) — the management REST API ``/api/*`` (app host only).
  * ``ui``    (P3) — the dashboard UI + static + WS feed ``/``, ``/d/*``,
                     ``/static/*``, ``/ws/*``, ``/sse/*``, ``/healthz`` (app host).

The bare apex (``https://<MOCK_DOMAIN>/`` with no subdomain), ``localhost``,
``127.0.0.1``, ``[::1]`` and ``<APP_HOST>`` all resolve to the UI plane (AC-6a).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional, Tuple

import config

# Path prefixes that, on the **app host**, belong to the management/UI planes and
# must never be served by the mock catch-all (AC-6).
_API_PREFIX = "/api"
_UI_EXACT = {"/", "/healthz", "/favicon.ico", "/robots.txt"}
_UI_PREFIXES = ("/static/", "/static", "/d/", "/d", "/ws/", "/ws", "/sse/", "/sse")


@dataclass(frozen=True)
class PlaneResult:
    plane: str                      # "mock" | "api" | "ui"
    token: Optional[str] = None     # endpoint token when plane == "mock"
    mock_path: Optional[str] = None  # the mock path handed to the engine (P1)


def _strip_port(host: str) -> str:
    """Drop a trailing ``:port`` (and brackets for bare IPv6 literals are kept)."""
    if not host:
        return ""
    host = host.strip()
    # IPv6 literal in brackets, optionally with a port: [::1]:8000
    if host.startswith("["):
        end = host.find("]")
        if end != -1:
            return host[: end + 1].lower()
        return host.lower()
    # host:port -> host (only split the LAST colon to be safe for bare names)
    if ":" in host:
        # A bare IPv6 without brackets would contain multiple colons; such a value
        # is not a valid Host header for our purposes, so only strip a single
        # trailing :digits port.
        head, _, tail = host.rpartition(":")
        if tail.isdigit():
            return head.lower()
    return host.lower()


def _strip_port_preserve_case(host: str) -> str:
    """Like ``_strip_port`` but does NOT lower-case — used to recover the
    case-sensitive endpoint token from a subdomain label (tokens use a mixed-case
    alphabet; DNS treats the label case-insensitively but our token does not, so
    we must preserve the label's original case)."""
    if not host:
        return ""
    host = host.strip()
    if host.startswith("["):
        end = host.find("]")
        return host[: end + 1] if end != -1 else host
    if ":" in host:
        head, _, tail = host.rpartition(":")
        if tail.isdigit():
            return head
    return host


def subdomain_of(host: str, mock_domain: str) -> Optional[str]:
    """Return the case-preserved ``<token>`` iff ``host == <token>.<mock_domain>``
    (single label), matching the domain suffix **case-insensitively**.

    Returns ``None`` for the bare apex, the app host, or any non-matching host.
    A multi-label subdomain (``a.b.<domain>``) is **not** a valid single token
    and returns ``None`` (it falls through to the UI plane).

    The token label's **case is preserved** because endpoint tokens use a
    mixed-case alphabet (``gen_token``); lower-casing it would break subdomain
    addressing (AC-4) vs the case-sensitive path-fallback (AC-5).
    """
    if not host or not mock_domain or "." not in mock_domain:
        return None
    raw = _strip_port_preserve_case(host)
    md = mock_domain.lower().strip()
    suffix = "." + md
    if not raw.lower().endswith(suffix):
        return None
    label = raw[: len(raw) - len(suffix)]
    if not label or "." in label:
        return None  # apex (empty) or multi-label -> not a mock token
    return label


def _path_fallback_token(path: str) -> Tuple[Optional[str], Optional[str]]:
    """Parse ``/e/<token>/<rest>`` → ``(token, mock_path="/<rest>")`` (arch §3.2).

    ``/e/<token>`` (no trailing path) → ``(token, "/")``. ``/e`` alone → no token.
    """
    # path is the URL path, already starting with "/".
    parts = path.split("/", 3)  # ['', 'e', '<token>', '<rest...>']
    if len(parts) < 3 or parts[1] != "e":
        return None, None
    token = parts[2]
    if not token:
        return None, None
    rest = parts[3] if len(parts) > 3 else ""
    mock_path = "/" + rest
    return token, mock_path


def resolve_plane(host: str, path: str) -> PlaneResult:
    """Resolve the plane for ``(host, path)`` (arch §3.1).

    Order matters: a wildcard mock host wins first (everything on it is mock),
    then the localhost path-fallback, then ``/api`` (management), then UI.
    """
    # P1a — wildcard mock host: EVERYTHING here is the mock's own path.
    sub = subdomain_of(host, config.MOCK_DOMAIN)
    if sub is not None and sub not in config.APP_HOSTS:
        return PlaneResult(plane="mock", token=sub, mock_path=path or "/")

    # P1b — localhost path-fallback /e/<token>/<rest> (any host).
    if path == "/e" or path == "/e/" or path.startswith("/e/"):
        token, mock_path = _path_fallback_token(path)
        if token is not None:
            return PlaneResult(plane="mock", token=token, mock_path=mock_path)
        # Malformed /e with no token -> UI 404 (not a mock).
        return PlaneResult(plane="ui")

    # From here we are on an app host (or the bare apex / unknown host → UI).
    # P2 — management API.
    if path == _API_PREFIX or path.startswith(_API_PREFIX + "/"):
        return PlaneResult(plane="api")

    # P3 — UI / static / ws / sse / health.
    if path in _UI_EXACT:
        return PlaneResult(plane="ui")
    for pref in _UI_PREFIXES:
        if path == pref or path.startswith(pref + "/") or path.startswith(pref):
            return PlaneResult(plane="ui")

    # Default app-host fallthrough → UI (the UI router returns 404 inside).
    return PlaneResult(plane="ui")

File: backend/src/planes.rs
//! Pure 3-plane resolution logic — 1:1 PORT of `app/planes.py` (arch §3.1/§3.2).
//!
//! Decides which of the three hard-isolated request planes a request belongs to,
//! purely from `Host` + `path` — no DB, no async — so it is trivially testable.
//! The router middleware (`router.rs`) applies it.
//!
//! Planes:
//! - `Mock` (P1) — the public wildcard mock surface `<token>.<MOCK_DOMAIN>/<path>`
//!   and the localhost path-fallback `/e/<token>/<path>`. On a mock host
//!   EVERYTHING (incl. `/api`, `/static`) is the mock's own path — management is
//!   unreachable there by construction.
//! - `Api` (P2) — the management REST API `/api/**` (app host only).
//! - `Ui` (P3) — dashboard + static + WS/SSE feed + health (app host).
//!
//! The bare apex, `localhost`, `127.0.0.1`, `[::1]`, `<APP_HOST>` all resolve to
//! the UI plane (AC-6a). The single load-bearing guarantee carried verbatim:
//! the mock catch-all can NEVER shadow P2/P3.

use std::collections::HashSet;

/// Path prefixes/exacts that, on the app host, belong to UI/management planes
/// and must never be served by the mock catch-all (AC-6).
const API_PREFIX: &str = "/api";
const UI_EXACT: [&str; 4] = ["/", "/healthz", "/favicon.ico", "/robots.txt"];
const UI_PREFIXES: [&str; 8] = [
    "/static/", "/static", "/d/", "/d", "/ws/", "/ws", "/sse/", "/sse",
];

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Plane {
    Mock,
    Api,
    Ui,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PlaneResult {
    pub plane: Plane,
    /// Endpoint token when `plane == Mock`.
    pub token: Option<String>,
    /// The mock path handed to the engine (P1).
    pub mock_path: Option<String>,
}

impl PlaneResult {
    fn ui() -> Self {
        PlaneResult {
            plane: Plane::Ui,
            token: None,
            mock_path: None,
        }
    }
    fn api() -> Self {
        PlaneResult {
            plane: Plane::Api,
            token: None,
            mock_path: None,
        }
    }
    fn mock(token: String, mock_path: String) -> Self {
        PlaneResult {
            plane: Plane::Mock,
            token: Some(token),
            mock_path: Some(mock_path),
        }
    }
}

/// Drop a trailing `:port`; preserve bracketed IPv6 literals — recovers the
/// case-sensitive endpoint
/// token from a subdomain label (tokens use a mixed-case alphabet).
fn strip_port_preserve_case(host: &str) -> String {
    let host = host.trim();
    if host.is_empty() {
        return String::new();
    }
    if let Some(rest) = host.strip_prefix('[') {
        return match rest.find(']') {
            Some(end) => host[..end + 2].to_string(),
            None => host.to_string(),
        };
    }
    if host.contains(':') {
        let (head, tail) = host.rsplit_once(':').unwrap();
        if !tail.is_empty() && tail.chars().all(|c| c.is_ascii_digit()) {
            return head.to_string();
        }
    }
    host.to_string()
}

/// Return the case-preserved `<token>` iff `host == <token>.<mock_domain>`
/// (single label), matching the domain suffix case-insensitively. None for the
/// bare apex, app host, multi-label subdomain, or non-matching host.
pub fn subdomain_of(host: &str, mock_domain: &str) -> Option<String> {
    if host.is_empty() || mock_domain.is_empty() || !mock_domain.contains('.') {
        return None;
    }
    let raw = strip_port_preserve_case(host);
    let md = mock_domain.trim().to_lowercase();
    let suffix = format!(".{md}");
    if !raw.to_lowercase().ends_with(&suffix) {
        return None;
    }
    let label = &raw[..raw.len() - suffix.len()];
    if label.is_empty() || label.contains('.') {
        return None; // apex (empty) or multi-label -> not a mock token
    }
    Some(label.to_string())
}

/// Parse `/e/<token>/<rest>` -> `(token, mock_path="/<rest>")`.
/// `/e/<token>` -> `(token, "/")`. `/e` alone -> no token.
fn path_fallback_token(path: &str) -> (Option<String>, Option<String>) {
    // Mirror Python `path.split("/", 3)` -> ['', 'e', '<token>', '<rest...>'].
    let parts: Vec<&str> = path.splitn(4, '/').collect();
    if parts.len() < 3 || parts[1] != "e" {
        return (None, None);
    }
    let token = parts[2];
    if token.is_empty() {
        return (None, None);
    }
    let rest = if parts.len() > 3 { parts[3] } else { "" };
    (Some(token.to_string()), Some(format!("/{rest}")))
}

/// Resolve the plane for `(host, path)` (arch §3.1). Order matters: a wildcard
/// mock host wins first (everything on it is mock), then the path-fallback,
/// then `/api`, then UI.
pub fn resolve_plane(
    host: &str,
    path: &str,
    mock_domain: &str,
    app_hosts: &HashSet<String>,
) -> PlaneResult {
    // P1a — wildcard mock host: EVERYTHING here is the mock's own path.
    if let Some(sub) = subdomain_of(host, mock_domain) {
        // A subdomain that collides with an app host stays UI (token==app host probe).
        if !app_hosts.contains(&sub) {
            let mock_path = if path.is_empty() {
                "/".to_string()
            } else {
                path.to_string()
            };
            return PlaneResult::mock(sub, mock_path);
        }
    }

    // P1b — path-fallback /e/<token>/<rest> (any host).
    if path == "/e" || path == "/e/" || path.starts_with("/e/") {
        let (token, mock_path) = path_fallback_token(path);
        if let (Some(t), Some(mp)) = (token, mock_path) {
            return PlaneResult::mock(t, mp);
        }
        // Malformed /e with no token -> UI 404 (not a mock).
        return PlaneResult::ui();
    }

    // App host (or bare apex / unknown host -> UI).
    // P2 — management API.
    if path == API_PREFIX || path.starts_with(&format!("{API_PREFIX}/")) {
        return PlaneResult::api();
    }

    // P3 — UI / static / ws / sse / health.
    if UI_EXACT.contains(&path) {
        return PlaneResult::ui();
    }
    for pref in UI_PREFIXES {
        // Preserve the Python's loose match: path == pref OR starts_with(pref+'/')
        // OR starts_with(pref).
        if path == pref || path.starts_with(&format!("{pref}/")) || path.starts_with(pref) {
            return PlaneResult::ui();
        }
    }

    // Default app-host fallthrough -> UI (the UI router returns 404 inside).
    PlaneResult::ui()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn app_hosts() -> HashSet<String> {
        ["localhost", "127.0.0.1", "[::1]", "app.local", "mock.local"]
            .iter()
            .map(|s| s.to_string())
            .collect()
    }

    fn resolve(host: &str, path: &str) -> PlaneResult {
        resolve_plane(host, path, "mock.local", &app_hosts())
    }

    #[test]
    fn wildcard_mock_host_captures_everything_incl_api_and_static() {
        // AC-4 / AC-6 / AC-S16: /api under a mock host stays P1, never P2.
        let r = resolve("aB3xZ.mock.local", "/api/endpoints");
        assert_eq!(r.plane, Plane::Mock);
        assert_eq!(r.token.as_deref(), Some("aB3xZ"));
        assert_eq!(r.mock_path.as_deref(), Some("/api/endpoints"));

        let r2 = resolve("aB3xZ.mock.local:8080", "/static/app.js");
        assert_eq!(r2.plane, Plane::Mock);
        assert_eq!(r2.token.as_deref(), Some("aB3xZ"));
    }

    #[test]
    fn subdomain_label_case_preserved() {
        // AC-4: token alphabet is mixed-case; the label case must survive.
        let r = resolve("MixedCase.MOCK.LOCAL", "/x");
        assert_eq!(r.token.as_deref(), Some("MixedCase"));
    }

    #[test]
    fn multi_label_subdomain_is_not_a_token() {
        // a.b.<domain> -> not a single token -> UI.
        let r = resolve("a.b.mock.local", "/x");
        assert_eq!(r.plane, Plane::Ui);
    }

    #[test]
    fn bare_apex_and_known_app_hosts_are_ui() {
        assert_eq!(resolve("mock.local", "/").plane, Plane::Ui);
        assert_eq!(resolve("localhost", "/api/endpoints").plane, Plane::Api);
        assert_eq!(resolve("127.0.0.1:8080", "/").plane, Plane::Ui);
        assert_eq!(resolve("[::1]:8080", "/healthz").plane, Plane::Ui);
        assert_eq!(resolve("app.local", "/d/abc").plane, Plane::Ui);
    }

    #[test]
    fn token_equal_to_app_host_does_not_become_mock() {
        // AC-S16 crafted probe: subdomain label collides with an app host name.
        let r = resolve("localhost.mock.local", "/api/endpoints");
        // 'localhost' is in APP_HOSTS -> not treated as a mock token -> /api is P2.
        assert_eq!(r.plane, Plane::Api);
    }

    #[test]
    fn path_fallback_e_token() {
        let r = resolve("localhost", "/e/aB3xZ/users/1");
        assert_eq!(r.plane, Plane::Mock);
        assert_eq!(r.token.as_deref(), Some("aB3xZ"));
        assert_eq!(r.mock_path.as_deref(), Some("/users/1"));

        let r2 = resolve("localhost", "/e/aB3xZ");
        assert_eq!(r2.plane, Plane::Mock);
        assert_eq!(r2.mock_path.as_deref(), Some("/"));

        // /e alone -> UI (malformed, not a mock).
        assert_eq!(resolve("localhost", "/e").plane, Plane::Ui);
        assert_eq!(resolve("localhost", "/e/").plane, Plane::Ui);
    }

    #[test]
    fn api_and_ui_classification_on_app_host() {
        assert_eq!(resolve("localhost", "/api").plane, Plane::Api);
        assert_eq!(
            resolve("localhost", "/api/endpoints/x/rules").plane,
            Plane::Api
        );
        assert_eq!(resolve("localhost", "/ws/tok").plane, Plane::Ui);
        assert_eq!(resolve("localhost", "/sse/tok").plane, Plane::Ui);
        assert_eq!(resolve("localhost", "/static/app.css").plane, Plane::Ui);
        // Unknown app-host path -> UI (SPA router 404s inside).
        assert_eq!(resolve("localhost", "/anything").plane, Plane::Ui);
    }

    #[test]
    fn percent_encoded_host_does_not_smuggle_into_mock() {
        // A percent-encoded label cannot match the literal mock-domain suffix.
        let r = resolve("%2e.mock.local", "/api/x");
        // '%2e' is a single label without a literal dot -> it IS a token here,
        // but /api is on the mock host so it stays P1 (never P2). Crucially it
        // does not reach the management plane.
        assert_eq!(r.plane, Plane::Mock);
    }

    #[test]
    fn path_fallback_only_when_mock_domain_dotless() {
        // MOCK_DOMAIN without a dot -> no wildcard mock; only /e/ fallback works.
        let hosts = app_hosts();
        let r = resolve_plane("anything.localhost", "/api/x", "localhost", &hosts);
        assert_eq!(r.plane, Plane::Api); // no wildcard -> /api is management
        let r2 = resolve_plane("anything.localhost", "/e/tok/x", "localhost", &hosts);
        assert_eq!(r2.plane, Plane::Mock);
        assert_eq!(r2.token.as_deref(), Some("tok"));
    }
}
